src/metabase/api/pulse.clj

```diff
@@ -117,6 +116,35 @@
   ;; return updated Pulse
   (pulse/retrieve-pulse id))
 
+(api/defendpoint GET "/form_input_refresh"
+  "Provides relevant configuration information and user choices for creating/updating Pulses, can block for quite a long
+  amount of time if the conversations-list or the users-list is huge."
+  []
+  (let [chan-types (-> channel-types
+                       (assoc-in [:slack :configured] (slack/slack-configured?))
+                       (assoc-in [:email :configured] (email/email-configured?)))]
+    {:channels (cond
+                 (when-let [segmented-user? (resolve 'metabase-enterprise.sandbox.api.util/segmented-user?)]
+                   (segmented-user?))
+                 (dissoc chan-types :slack)
+
+                 ;; no Slack integration, so we are g2g
+                 (not (get-in chan-types [:slack :configured]))
+                 chan-types
+
+                 ;; if we have Slack enabled, block until we gather the entire list of conversations + users (can be
+                 ;; quite a while).
+                 :else
+                 (try
+                   (let [[conversations users] (map deref [(future (slack/conversations-list))
+                                                           (future (slack/users-list))])
+                         _                     (slack/slack-cache {:conversations conversations :users users})
+                         slack-channels        (for [channel conversations] (str \# (:name channel)))
+                         slack-users           (for [user users] (str \@ (:name user)))]
+                     (assoc-in chan-types [:slack :fields 0 :options] (concat slack-channels slack-users)))
+                   (catch Throwable e
+                     (assoc-in chan-types [:slack :error] (.getMessage e)))))}))
```

Comment:
Some thoughts:
1. Since `conversations-list` and `users-list` both return lazy sequences you're not actually running these on separate threads the way you think. You're fetching the first page for each on separate threads, but these aren't actually ultimately realized until you call `slack/slack-cache` below, at which point the lazy sequences get realized on a single thread.
2. Since we only ever use the `@name` or `#name` info we should calculate that here and store _that_ in the cache rather than storing the entire enormous Slack API response in the cache and then calculating the same stuff when we pull the cached results out. 
3. You should be converting channels/users to `#name` or `@name` immediately as the lazy seq is realized rather than realizing the whole thing and _then_ converting them -- as @noahmoss mentioned for a huge Slack instance with millions of Users that's going to make a huge difference in terms of the peak amount of memory this uses (1 million giant maps is a lot bigger than 1 million strings)

Here's one way you might consider reworking it.

The basic idea is to fully realize `(slack/conversations)` and `(slack/users-list)` on separate threads. We transduce those and convert them to `#name` or `@name` immediately on the respective separate threads, then concat those together with `mapcat` (this itself is a lazy sequence but that ultimately won't make a difference here since we're fully realizing stuff in the background threads). 

You can store this concatenated list in the cache rather than the entire API response. 

Let me know if this makes sense.

```suggestion
                 (try
                   (let [channels (mapcat deref [(future (mapv #(str \# (:name %))) (slack/conversations-list))
                                                 (future (mapv #(str \@ (:name %))) (slack/users-list))])]
                     ;; update the cached channels
                     (slack/slack-cache channels)
                     (assoc-in chan-types [:slack :fields 0 :options] channels))
                   (catch Throwable e
                     (assoc-in chan-types [:slack :error] (.getMessage e)))))}))
```

Reply:
I guess my comments about this not really being truly multithreaded aren't 100% true since you changed the underlying functions to fully realized everything (by wrapping in `(vec ...)`). But I think that's overall a bad change since we'd be forced to realize the entire API response and couldn't discard unneeded bits as we go. I'll add some more comments about why to that part of the diff